tests/test_dom0_and_templates_correct_key.py

```diff
@@ -0,0 +1,248 @@
+            # First line after header should be without "."
+            if line.strip() == ".":
+                key_lines.append(" ")
+            else:
+                key_lines.append(line.strip())
+            if "END PGP PUBLIC KEY BLOCK" in line:
+                break
+        elif line.strip() == "Signed-By:":
+            capturing = True  # Start capturing from the next line
+
+    if not key_lines or "END PGP PUBLIC KEY BLOCK" not in key_lines[-1]:
+        raise ValueError("No complete PGP key found")
+
+    with tempfile.NamedTemporaryFile(
+        delete=False, suffix=".asc", mode="w", encoding="utf-8"
+    ) as tmp_file:
+        tmp_file.write("\n".join(key_lines) + "\n")
+
+    return tmp_file.name
+
+
+### BEGIN TESTS ###
+
+# Get config, needed for test
+ENV, CONFIG = _get_expected_config()
+
+
+def test_dom0_has_keyring_package():
+    # Prod keyring is always installed
+    assert _is_installed("securedrop-workstation-keyring")
+
+    # TODO: remove this check when config.json does not specify
+    # "environment" anymore.
+    # Until then, the order matters; a dev setup may also have
+    # a staging package, but not vice-versa.
+    if ENV == "dev":
+        assert _is_installed("securedrop-workstation-keyring-dev")
+    elif ENV == "staging":
+        assert _is_installed("securedrop-workstation-keyring-staging")
+
+
+def test_rpm_repo_config():
+    repo = CONFIG["repo_file_name"]
+    baseurl = CONFIG["yum_repo_url"]
+    repo_file = f"/etc/yum.repos.d/{repo}"
+    wanted_lines = [
+        "[securedrop-workstation-dom0]",
+        "gpgcheck=1",
+        "skip_if_unavailable=False",
+        "gpgkey=file://{}".format(CONFIG.get("signing_key")),
+        "enabled=1",
+        f"baseurl={baseurl}",
+        "name=SecureDrop Workstation Qubes dom0 repo",
+    ]
+    with open(repo_file) as f:
+        found_lines = [x.strip() for x in f.readlines()]
+
+    assert found_lines == wanted_lines
+
+
+def test_key_dom0():
+    """Check the key in /etc/pki/rpm-gpg/, used for dom0 updates."""
+    _read_and_validate_key(
+        CONFIG["signing_key"],
+        CONFIG["signing_key_fpr"],
+        CONFIG["signing_key_uid"],
+        CONFIG["signing_key_exp"],
+    )
+
+
+def test_key_apt_sources():
+    """Check the key in the apt_sources file used for provisioning Debian-based templates."""
+    app = Qubes()
+    for template in ["small", "large"]:
+        vm = app.domains[f"sd-{template}-{DEBIAN_DIST}-template"]
+        apt_sources_filename = CONFIG["apt_sources"]
+        stdout, _ = vm.run(f"cat {apt_sources_filename}")
+        sources = stdout.decode()
+
+        tmp_keyfile = _parse_key_from_sources(sources)
+
+        # Know which template we failed on
+        try:
+            _read_and_validate_key(
+                tmp_keyfile,
+                CONFIG["signing_key_fpr"],
+                CONFIG["signing_key_uid"],
+                CONFIG["signing_key_exp"],
+            )
+        except ValueError as e:
```

Comment:
Above I suggested using assertions in `_read_and_validate_key` so then we don't need to catch exception and then immediately fail. Plus if we use parameterization then the test name will also have the VM name.